ArcadiaTeamsBot/Dialogs/NewRequestDialog.cs

```diff
@@ -1,63 +1,271 @@
+    using Newtonsoft.Json.Linq;
+
+    internal class NewRequestDialog : ComponentDialog
     {
-        private const string Back = "Back";
+        private const string Submit = "Submit";
+        private const string Cancel = "Cancel";
+        private const string Yes = "Yes";
+        private const string No = "No";
+        private const string Username = "ekaterina.kuznetsova@arcadia.spb.ru";
+        private readonly IRequestTypeUIFactory requestTypeUIFactory;
+        private readonly IMediator mediator;
 
-        public NewRequestDialog() : base(nameof(NewRequestDialog))
+        public NewRequestDialog(IMediator mediator, IRequestTypeUIFactory requestTypeUIFactory) : base(nameof(NewRequestDialog))
         {
+            this.mediator = mediator;
+            this.requestTypeUIFactory = requestTypeUIFactory;
+
             this.AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
             {
-                InfoStep,
-                EndStep,
+                this.InputStep,
+                this.ChoiceStep,
+                this.EndStep
             }));
 
             this.AddDialog(new TextPrompt(nameof(TextPrompt)));
             this.InitialDialogId = nameof(WaterfallDialog);
+            this.AddDialog(new TextPrompt("askForInput", this.ValidateForm));
         }
 
-        private static async Task<DialogTurnResult> InfoStep(WaterfallStepContext stepContext, CancellationToken cancellationToken)
+        private async Task<bool> ValidateForm(PromptValidatorContext<string> promptContext, CancellationToken cancellationToken)
         {
-            var attachments = new[]
+            var formData = (JObject)promptContext.Context.Activity.Value;
+            if (formData == null)
+            {
+                return false;
+            }
+
+            if (formData["Button"].ToString() == Cancel)
+            {
+                return true;
+            }
+
+            var additionalFieldsValues = new List<string>();
+            var additionalFieldsData = (formData).Properties()
+                .Where(p => p.Value.Type == JTokenType.String).ToList();
+
+            for (var i = 6; i < additionalFieldsData.Count; i++)
             {
-                GetInfoCard().ToAttachment()
+                additionalFieldsValues.Add(additionalFieldsData[i].Last.ToString());
             };
-            return await stepContext.PromptAsync(nameof(TextPrompt),
-                new PromptOptions
+
+            if (string.IsNullOrEmpty(formData["Title"].ToString())
+                || string.IsNullOrEmpty(formData["Description"].ToString())
+                || string.IsNullOrEmpty(formData["ExecutionDate"].ToString())
+                || additionalFieldsValues.Any(string.IsNullOrEmpty))
+            {
+                return false;
+            }
+
+            var createRequestTypeFields =
+                (await this.mediator.Send(new GetServiceDeskRequestTypesQuery(), cancellationToken))
```

Comment:
Move getting of request type fields to separate method.

Reply:
Moved